cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
```

Comment:
At the same time it should extend NotFoundException? To allow generic HTTP error handling?

Reply:
We don't extend that one from javax.ws. For CDAP handlers, we check for class name to have `NotFoundException` in it.

Reply:
I don't think we should throw NotFoundException here. Since now we will support custom resource tagging we cannot check whether the resource exists or not. We can return an empty list in most cases if we don't find any metadata for the given key. How about this behavior?